src/helper/OrderHelper.sol

```diff
@@ -0,0 +1,218 @@
+    IERC20 public oldCollateral;
+    IERC20 public oldCollateralAToken;
+    uint256 public oldCollateralAmount;
+    IERC20 public newCollateral;
+    IERC20 public newCollateralAToken;
+    uint256 public minSupplyAmount;
+    uint32 public validTo;
+    uint256 public flashloanFee;
+    address public flashloanPayee;
+    address public factory;
+    uint256 transient preHookCalled;
+
+    function initialize(
+        address _owner,
+        address _tracker,
+        address _oldCollateral,
+        uint256 _oldCollateralAmount,
+        address _newCollateral,
+        uint256 _minSupplyAmount,
+        uint32 _validTo,
+        uint256 _flashloanFee,
+        address _flashloanPayee,
+        address _factory
+    ) external initializer {
+        // TODO: check the other params?
+        if (_owner == address(0)) {
+            revert OrderHelperError.BadParameters();
+        }
+
+        if (_validTo < block.timestamp) {
+            revert OrderHelperError.BadParameters();
+        }
+
+        owner = _owner;
+        tracker = _tracker;
+        oldCollateral = IERC20(_oldCollateral);
+        newCollateral = IERC20(_newCollateral);
+        oldCollateralAmount = _oldCollateralAmount;
+        minSupplyAmount = _minSupplyAmount;
+        validTo = _validTo;
+        flashloanFee = _flashloanFee;
+        flashloanPayee = _flashloanPayee;
+        factory = _factory;
+
+        AAVE_LENDING_POOL = IOrderFactory(factory).AAVE_LENDING_POOL();
+        oldCollateralAToken = IERC20(IAavePool(AAVE_LENDING_POOL).getReserveAToken(address(_oldCollateral)));
+        newCollateralAToken = IERC20(IAavePool(AAVE_LENDING_POOL).getReserveAToken(address(_newCollateral)));
+
+        // Approve the _oldCollateral AToken for the swap
+        IERC20(oldCollateralAToken).forceApprove(ISettlement(SETTLEMENT).vaultRelayer(), type(uint256).max);
+
+        // The system will pull the old collateral to payback the flash loan
+        IERC20(oldCollateral).forceApprove(tracker, type(uint256).max);
+    }
+
+    // Prehook will take care of depositing the flash loan amount into aave
+    function preHook() external {
+        if (preHookCalled != 0) {
+            revert OrderHelperError.PreHookAlreadyCalled();
+        }
+        preHookCalled = 1;
+
+        // It should be the same amount, but someone can dust
+        if (oldCollateral.balanceOf(address(this)) < oldCollateralAmount) {
+            revert OrderHelperError.NotEnoughOldCollateral();
+        }
+
+        // Approve and deposit the old collateral into aave
+        IERC20(oldCollateral).forceApprove(AAVE_LENDING_POOL, oldCollateralAmount);
+        IAavePool(AAVE_LENDING_POOL).supply(address(oldCollateral), oldCollateralAmount, address(this), 0);
+    }
+
+    function isValidSignature(bytes32 _orderHash, bytes calldata _signature) external view returns (bytes4) {
+        if (preHookCalled != 1) {
+            revert OrderHelperError.PreHookNotCalled();
+        }
```

Comment:
I don't think this check is necessary. If the preHook wasn't called, the contract wouldn't have sufficient sell token balance and so the settlement will fail when it tries to transfer the sell amount into the settlement contract.

Not having this assert should make order placement possible in the current backend code without compromising security.